File: 2017/6.py
mems = "14	0	15	12	11	11	3	5	1	6	8	4	9	1	8	4"

part = 2
banks = list(map(int, mems.split()))
old = [banks]
cycles = [0]
size = len(banks)
steps = 0
while True:
    start = banks.index(max(banks))
    new_bank = list(banks)
    offset = start
    for i in range(banks[start]):
        new_bank[start] -= 1
        offset += 1
        offset %= size
        new_bank[offset] += 1

    steps += 1
    if new_bank not in old:
        old.append(new_bank)
        cycles.append(steps)
    else:
        break
    banks = new_bank

if part == 1:
    print ("🎄Part 1: {}".format(steps))
else:
    idx = old.index(new_bank)
    print ("🎁🎄Part 2: {}".format(steps-cycles[idx]))

File: 2019/6.jl
function allocate(part=1)
    mems = "14	0	15	12	11	11	3	5	1	6	8	4	9	1	8	4"

    banks = parse.(Int, split(mems))
    old = [banks]
    cycles = [0]
    size = length(banks)
    steps = 0
    last_new_bank = []
    while true
        start = findfirst(x->x==maximum(banks), banks)
        new_bank = copy(banks)
        offset = start
        for i in 1:banks[start]
            new_bank[start] -= 1
            offset += 1
            offset = offset%size==0 ? size : offset%size
            new_bank[offset] += 1
        end

        steps += 1
        if !(new_bank in old)
            push!(old, new_bank)
            push!(cycles, steps)
        else
            last_new_bank = new_bank
            break
        end
        banks = new_bank
    end

    if part == 1
        println("🎄 $(steps)")
    else
        idx = findfirst(x->x==last_new_bank, old)
        println("🎁🎄 $(steps-cycles[idx])")
    end
end

allocate()
allocate(2)